Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2.

pearu opened the pull request:
Replaced by https://github.com/pytorch/pytorch/pull/75821

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75585
* __->__ #75596
* #75962
* #75961

pearu committed:
Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -44,20 +57,24 @@ SparseCsrTensorImpl::SparseCsrTensorImpl(
+      layout_(layout) {
   // https://pytorch.org/blog/pytorch-feature-classification-changes/#beta
-  TORCH_WARN_ONCE("Sparse CSR tensor support is in beta state.");
+  TORCH_WARN_ONCE("Sparse ", SparseCsrTensorLayoutToSTRING(layout_), " tensor support is in beta state.");
```
We can also extend this by telling the user to consider opening a feature request if there is missing functionality with a link to the page: https://github.com/pytorch/pytorch/issues

pearu replied:
Extended the warning message with "If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues."

pearu replied:
Resolved via https://github.com/pytorch/pytorch/pull/75594

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -19,10 +19,23 @@ DeviceType SparseCsrTensorSetToDeviceType(DispatchKeySet key_set) {
   }
 }
+
+std::string SparseCsrTensorLayoutToSTRING(Layout layout) {
```
The capitalization of "string" is nonstandard (see other comparable toString functions), let's go with "String" instead please.

pearu replied:
On the second though, introducing a `string SparseCsrTensorImpl::layoutToString()` method is more useful.

pearu replied:
Resolved via layoutToString implementation in https://github.com/pytorch/pytorch/pull/75865

cpuhrsch commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -68,8 +85,11 @@ const char* SparseCsrTensorImpl::tensorimpl_type_name() const {
-  auto rows = size[size.size() - 2];
-  auto cols = size[size.size() - 1];
+  TORCH_CHECK((layout_ == kSparseCsr || layout_ == kSparseBsr), "resize_: layout ", layout_, " is not yet supported");  // TODO
+  auto row_index = size.size() - ((layout_ == kSparseCsr || layout_ == kSparseBsr) ? 2 : 1);
```
an "SparseCsrTensorImpl::is_transposed" method might be convenient to wrap ```!(layout_ == kSparseCsr || layout_ == kSparseBsr)```. I see 7 invocations of this pattern in this PR.

cpuhrsch replied:
@pearu - that works too. For now it's just a means of reducing the amount of code.

pearu replied:
Resolved via different layout specific dispatch macros in https://github.com/pytorch/pytorch/pull/75865

cpuhrsch commented on aten/src/ATen/SparseCsrTensorUtils.h:
```diff
@@ -11,9 +11,17 @@ namespace sparse_csr {
+  default:
+    AT_ASSERTM(
+               false,
+               "_internal_get_SparseCsrTensorImpl: not a sparse CSR|CSC|BSR|BSC tensor");
```
nit: You could also print the layout that was passed.

pearu replied:
Resolved via https://github.com/pytorch/pytorch/pull/75865

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5351,7 +5351,7 @@
 - func: _validate_sparse_coo_tensor_args(Tensor indices, Tensor values, int[] size) -> ()
 
-- func: _validate_sparse_csr_tensor_args(Tensor crow_indices, Tensor col_indices, Tensor values, int[] size) -> ()
+- func: _validate_sparse_csr_tensor_args(Tensor crow_indices, Tensor col_indices, Tensor values, int[] size, Layout? layout=None) -> ()
```
Can you make this a keyword-only argument to avoid BC issues?

pearu replied:
Resolved via https://github.com/pytorch/pytorch/pull/75946

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -268,30 +332,46 @@ Tensor empty_sparse_csr(
-  crow_indices_size.push_back(rows + 1);
-  auto col_indices_values_size = DimVector(size.slice(0, size.size() - 2));
-  col_indices_values_size.push_back(nnz);
+  Layout layout_ = layout.value_or((size.size() == 2 ? kSparseCsr : kSparseBsr));  // TODO: this is wrong..
```
I think it's fair to just assume kSparseCsr unless the layout was explicitly passed.

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -357,13 +464,22 @@ bool _is_same_size_as_sparse_csr(
-      " for self, and ",
-      src.layout(),
-      " for src");
+  switch(self.layout()) {
```
If we had an `AT_DISPATCH_LAYOUT_SPARSE_COMPRESSED`, akin to the functions in [Dispatch.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h), the entirety of
```
switch(self.layout()) {
  case kSparseCsr:
  case kSparseCsc:
  case kSparseBsr:
  case kSparseBsc:
    TORCH_CHECK(
                src.layout() == self.layout(),
                "resize_as_sparse_csr_: layout for self and src must be match but got ",
                self.layout(),
                " for self, and ",
                src.layout(),
                " for src");
    break;
  default:
    TORCH_CHECK(false, "resize_as_sparse_csr_: layout ", self.layout(), " is not supported");
  }
```

 could be written as

```
AT_DISPATCH_LAYOUT_SPARSE_COMPRESSED(self.layout(), "resize_as_sparse_csr_", [&]() {
    TORCH_CHECK(
        src.layout() == self.layout(),
        "resize_as_sparse_csr_: layout for self and src must be match but got ",
         self.layout(),
         " for self, and ",
         src.layout(),
         " for src");
});
```

I think that'd be a useful macro in general. What do you think?

pearu replied:
This makes sense: it will save/unify us typing `switch-cases` constructs and the `default` case handling.

pearu replied:
Resolved via https://github.com/pytorch/pytorch/pull/75865

cpuhrsch commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2028,7 +2028,7 @@ def random_sparse_csr(n_rows, n_cols, nnz):
         return torch.sparse_csr_tensor(crow_indices,
                                        col_indices,
-                                       values, size=size, dtype=dtype, device=device)
+                                       values, size=size, dtype=dtype, layout=torch.sparse_csr, device=device)
```
You shouldn't need this change, but I guess it doesn't hurt

cpuhrsch commented:
Generally this seems fine for me (see some comments inline). However, after I accepted it I'd like another core developer to review this as well, because it touches a lot of parts of PyTorch.

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -186,79 +234,156 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
-    c10::optional<Device> device,
-    c10::optional<bool> pin_memory) {
-
+template <Layout expected_layout>
+Tensor _sparse_compressed_tensor_unsafe_template(const Tensor& compressed_indices,
+                                                 const Tensor& plain_indices,
+                                                 const Tensor& values,
+                                                 IntArrayRef size,
+                                                 c10::optional<ScalarType> dtype,
+                                                 c10::optional<Layout> layout,
+                                                 c10::optional<Device> device,
+                                                 c10::optional<bool> pin_memory) {
+  Layout layout_ = layout.value_or(expected_layout);
+  if (expected_layout == kDummyLayout) {
+    TORCH_CHECK(layout_ != kDummyLayout, "expected layout SparseCsr, SparseCsc, SparseBsr, or SparseBsc but got none");
+  } else {
+    TORCH_CHECK(layout_ == expected_layout, "expected layout ", expected_layout, " but got ", layout_);
+  }
```
@cpuhrsch ^

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl and sparse CSC BSR BSC layouts, 2."

Fixes https://github.com/pytorch/pytorch/issues/74916

Replaces https://github.com/pytorch/pytorch/pull/75023, https://github.com/pytorch/pytorch/pull/74546, and https://github.com/pytorch/pytorch/pull/74975.




[ghstack-poisoned]

pearu commented:
Replaced by https://github.com/pytorch/pytorch/pull/75821
